## 微信素材管理
1. 临时素材
2. 永久素材
3. 以图片为例

In [3]:
import time
import requests, urllib2
import poster.encode
from poster.streaminghttp import register_openers

In [26]:
class Basic:
    # 获取access_token
    def __init__(self):        
        self.__accessToken = ''        
        self.__leftTime = 0
        self.url = "https://api.weixin.qq.com/cgi-bin/token?grant_type=" "client_credential&appid={aid}&secret={sid}"
        
    def __real_get_access_token(self):
        urlResp = requests.get(url=self.url.format(aid="wx5b9a64bcdb671577", 
                                                   sid="cffd473cf2cc2219cca7fef0b5eb871a")).json()
        # app Secret id需要通过忘记重新获取，正常状态是隐藏的
        print(urlResp)
        self.__accessToken = urlResp['access_token']
        self.__leftTime = urlResp['expires_in']
        
    def get_access_token(self): 
        if self.__leftTime < 10:
            self.__real_get_access_token()
            return self.__accessToken
        
    def run(self):
        while(True):
            if self.__leftTime > 10:
                time.sleep(2)
                self.__leftTime -= 2
            else:
                self.__real_get_access_token()

In [28]:
class Media(object):
    def __init__(self):
        register_openers()

    def upload(self, accessToken, filePath, mediaType):
        openFile = open(filePath, "rb")
        media_name = "dog and fish"
        param = {'media': openFile}
        # param.update({"articles": [{"title": media_name,
        #                             "thumb_media_id": abs(hash(media_name)),
        #                             "show_cover_pic": 1,
        #                             "content": "rand",
        #                             "content_source_url": "http://image.baidu.com"}]})
        postData, postHeaders = poster.encode.multipart_encode(param)
        # postUrl = "https://api.weixin.qq.com/cgi-bin/media/upload?access_token=%s&type=%s" % (accessToken, mediaType)
        postUrl = "https://api.weixin.qq.com/cgi-bin/material/add_material?access_token=%s&type=%s" \
                  % (accessToken, mediaType) # 永久素材获取url
        request = urllib2.Request(postUrl, postData, postHeaders)
        urlResp = urllib2.urlopen(request)
        print urlResp.read()


if __name__ == '__main__':
    accessToken = Basic().get_access_token()
    filePath = "/Users/fanzfeng/Documents/picture/yu.jpg"
    mediaType = "image"
    myMedia = Media()
    myMedia.upload(accessToken, filePath, mediaType)

{u'access_token': u'17_7TnPlPOE1jE-5qmhCbQ4Vy5fp9YpCSl-YwNBgvKmaoRkkKFIRuPb-WAVmnwfvzLP2HPvVXB06vGjT8KVSutYNodHmpshwUWTFrfwNVO8fzgKZAxOGgxvdLQhxLyBKcaZrdHUXKAZUo-vWN7uWGNdAGAPZG', u'expires_in': 7200}


{"errcode":48001,"errmsg":"api unauthorized hint: [tSC8LA04508055!]"}


`上传永久素材报错{"errcode":48001,"errmsg":"api unauthorized hint: [tSC8LA04508055!]"}，意思是微信上传永久素材需要是企业号，需要认证的`